pkg/controller/job/job_controller.go

```diff
@@ -0,0 +1,452 @@
+	queue *workqueue.Type
+}
+
+func NewJobManager(kubeClient client.Interface) *JobManager {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(kubeClient.Events(""))
+
+	jm := &JobManager{
+		kubeClient: kubeClient,
+		podControl: controller.RealPodControl{
+			KubeClient: kubeClient,
+			Recorder:   eventBroadcaster.NewRecorder(api.EventSource{Component: "job"}),
+		},
+		expectations: controller.NewControllerExpectations(),
+		queue:        workqueue.New(),
+	}
+
+	jm.jobStore.Store, jm.jobController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return jm.kubeClient.Experimental().Jobs(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return jm.kubeClient.Experimental().Jobs(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&experimental.Job{},
+		replicationcontroller.FullControllerResyncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc: jm.enqueueController,
+			UpdateFunc: func(old, cur interface{}) {
+				job := cur.(*experimental.Job)
+				for _, c := range job.Status.Conditions {
+					if c.Type == experimental.JobComplete && c.Status == api.ConditionTrue {
+						return
+					}
+				}
+				jm.enqueueController(cur)
+			},
+			DeleteFunc: jm.enqueueController,
+		},
+	)
+
+	jm.podStore.Store, jm.podController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return jm.kubeClient.Pods(api.NamespaceAll).List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(rv string) (watch.Interface, error) {
+				return jm.kubeClient.Pods(api.NamespaceAll).Watch(labels.Everything(), fields.Everything(), rv)
+			},
+		},
+		&api.Pod{},
+		replicationcontroller.PodRelistPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    jm.addPod,
+			UpdateFunc: jm.updatePod,
+			DeleteFunc: jm.deletePod,
+		},
+	)
+
+	jm.updateHandler = jm.updateJob
+	jm.syncHandler = jm.syncJob
+	jm.podStoreSynced = jm.podController.HasSynced
+	return jm
+}
+
+// Run the main goroutine responsible for watching and syncing jobs.
+func (jm *JobManager) Run(workers int, stopCh <-chan struct{}) {
+	defer util.HandleCrash()
+	go jm.jobController.Run(stopCh)
+	go jm.podController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go util.Until(jm.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down Job Manager")
+	jm.queue.ShutDown()
+}
+
+// getPodJob returns the job managing the given pod.
+func (jm *JobManager) getPodJob(pod *api.Pod) *experimental.Job {
+	jobs, err := jm.jobStore.GetPodJobs(pod)
+	if err != nil {
+		glog.V(4).Infof("No jobs found for pod %v, job manager will avoid syncing", pod.Name)
+		return nil
+	}
+	// TODO: add sorting and rethink the overlaping controllers, internally and with RCs
+	return &jobs[0]
```

Comment:
If there are more than one jobs for a pod, this is non-deterministic. This needs a sort by creation time stamp. It's a user/configuration error but this also needs a log statement.